source/custom/nimxlcterminaladaptor_restricted_service.custom.cpp

```diff
@@ -0,0 +1,80 @@
+#include <iostream>
+#include <vector>
+#include <server/converters.h>
+#include <nierr_Status.h>
+#include <nimxlcterminaladaptor_restricted.pb.h>
+#include <nimxlcterminaladaptor_restricted/nimxlcterminaladaptor_restricted_service.h>
+#include <custom/nimxlcterminaladaptor_restricted_converters.h>
+
+namespace nimxlcterminaladaptor_restricted_grpc {
+  
+  using nidevice_grpc::converters::allocate_output_storage;
+  using nidevice_grpc::converters::convert_to_grpc;
+  using nimxlcterminaladaptor_restricted_grpc::NIErrStatus;
+  
+::grpc::Status NimxlcTerminalAdaptorRestrictedService::GetDeviceContainer(::grpc::ServerContext* context, const GetDeviceContainerRequest* request, GetDeviceContainerResponse* response)
+{
+    if (context->IsCancelled()) {
+        return ::grpc::Status::CANCELLED;
+    }
+    try {
+        auto grpc_session = request->session();
+        nimxlc_Session session = session_repository_->access_session(grpc_session.id(), grpc_session.name());
+
+        nierr_Status status;
+        status.code = 0;
+        status.capacity = 0;
+        status.json = "";
+        nimxlc_DeviceContainer deviceContainer = library_->getDeviceContainer(session, &status);
+        if (nierr_Status_isNotFatal(&status))
+        {
+            nimxlc_DeviceIterator deviceIterator = library_->DeviceContainer_begin(deviceContainer);
+            while (!library_->DeviceIterator_isEnd(deviceContainer, deviceIterator) && nierr_Status_isNotFatal(&status))
+            {
+                auto device = response->add_container_out();
```

Comment:
If adding data to the response object throws an exception, the server will leak memory. Please consider using RAII to call the cleanup functions.

Reply:
Added DeviceContainerPtr and TerminalContainerPtr objects that call the appropriate cleanup functions in the destructor.